JSONRPC/JSONRPC.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.IO;
+using System.Text;
+
+using Newtonsoft.Json.Linq;
+
+public class JSONRPC {
+	private string _endpoint;
+
+	// Basic authentication
+	private string _username;
+	private string _password;
+
+	public void setEndpoint(string endpoint, string username = null, string password = null) {
+		_endpoint = endpoint;
+		_username = username;
+		_password = password;
+	}
+
+	public void call(string methodName, JObject parameters, Dictionary<string, string> headers, Action<Exception, JObject> cb) {
+		// Make sure the endpoint is set
+		if (string.IsNullOrEmpty(_endpoint)) {
+			cb(new Exception("Endpoint has not been set"), null);
+			return;
+		}
+
+		// Setup JSON request object
+		JObject requestObject = new JObject ();
+		requestObject.Add("id", new JValue(1));
+		requestObject.Add("jsonrpc", new JValue("2.0"));
+		requestObject.Add("method", new JValue(methodName));
+		requestObject.Add("params", parameters);
+
+		// Serialize JSON request object into string
+		string postData;
+		try {
+			postData = requestObject.ToString ();
+		} catch (Exception serializeError) {
+			cb(serializeError, null);
+			return;
+		}
+
+		// Make a copy of the provided headers and add additional required headers
+		Dictionary<string, string> _headers = new Dictionary<string, string>(headers);
+		if (_username != null && _password != null) {
+			string authInfo = _username + ":" + _password;
+			string encodedAuth = Convert.ToBase64String(Encoding.Default.GetBytes(authInfo));
+			_headers.Add("Authorization", "Basic " + encodedAuth);
+		}
+
+		// Send HTTP post to JSON rpc endpoint
+		HTTPRequest.Post (_endpoint, "application/json", postData, _headers, (Exception requestError, string responseString) => {
+			// Deserialize the JSON response
+			JObject responseObject;
+			try {
+				responseObject = JObject.Parse(responseString);
```

Comment:
That's fine in and of itself, but I mean to say that if this library here is supposed to deal with JSONRPC, it shouldn't limit itself to serializing the request, but also with deserializing the response and only return the relevant bits, am I wrong? I understand one reason for the current state of it is the events being returned with the response (which is a bit of a hack on top of JSONRPC), but it would be nice if we could figure out an elegant way to do response parsing inside this library, rather than outside.